[MRG+1] Merge PresortBestSplitter and BestSplitter

jmschrei opened the pull request:
This pull request merges PresortBestSplitter into BestSplitter, without loss of functionality. All tree based estimators (Decision Trees, Random Forests, Extra Trees, and Gradient Boosting) have an optional parameter "presort," which is default false for all except Gradient Boosting.

In addition to allowing other estimators to turn on presorting on smaller data sets, it allows Gradient Boosting to turn off presorting on large datasets. A good default for this switch needs to be found, as it currently is done manually. Gradient Boosting can now also work on sparse data by turning off the presorting option.

Here is a checklist of what needs to be done (much more manageable than my last one):
- [x] Merge PresortBestSplitter into BestSplitter
- [x] Write unit tests to test sparse data, presorting options for RF and DT, and non-presorting for GBTs 
- [x] Find a good default switch for GBT to switch from presorting to non-presorting

As a side note, I'm getting some pretty high variance running this benchmark, with sometimes my branch being faster and sometimes slower, by up to 50s. Can someone else run the benchmarks a few times and see what they get?

@arjoly @pprett @glouppe 

```
MASTER
Classification performance:
===========================
Classifier   train-time test-time error-rate
--------------------------------------------
RandomForest  32.1754s   0.4057s     0.0330  
ExtraTrees    39.5981s   0.5889s     0.0372  
CART          12.1820s   0.0209s     0.0424  
GBRT         584.7982s   0.4412s     0.1777

BRANCH
Classification performance:
===========================
Classifier   train-time test-time error-rate
--------------------------------------------
RandomForest  36.0459s   0.3898s     0.0330  
ExtraTrees    35.2664s   0.5764s     0.0372  
CART          15.3547s   0.2031s     0.0424  
GBRT         557.0054s   0.3247s     0.1777  
```

jmschrei committed:
ENH gbt sparse support

glouppe commented on sklearn/ensemble/forest.py:
```diff
@@ -251,6 +255,11 @@ def fit(self, X, y, sample_weight=None):
         random_state = check_random_state(self.random_state)
 
+        X_idx_sorted = None
+        if self.presort == True:
```
`if self.presort` is more pythonic

jmschrei commented:
Ran some time tests with and without presorting.

```
max_depth = 3

with presorting
spambase   0.91     0.11
Gaussian   0.853    8.51
mnist      0.837  191.26
covtypes   0.708  111.45

without presorting
spambase   0.91     0.14
Gaussian   0.853   20.29
mnist      0.837  263.88
covtypes   0.708  134.32

max_depth = 6
with presorting
spambase   0.925 0.29
Gaussian   0.911 22.286
mnist      0.912 718.27
covtypes   0.774 306.24

without presorting
spambase   0.926 0.284
Gaussian   0.911 38.73
mnist      0.913 549.19
covtypes   0.774 249.28
```

So it looks like some function of max_depth, n_samples, and n_features corresponds to training time. Do you think it's worth finding a good switch, or mentioning it in the docs and leaving it to users to determine if they want to presort or not?

glouppe commented:
> So it looks like some function of max_depth, n_samples, and n_features corresponds to training time. Do you think it's worth finding a good switch, or mentioning it in the docs and leaving it to users to determine if they want to presort or not?

I would leave that to the user and only enable presort for boosting. Let's keep things simple :)

jmschrei commented:
I have added unit tests for Gradient Boosting and Forests to ensure that presorting returns the same results as not using presorting. I have also added unit tests for Gradient Boosting testing sparse inputs, and ensuring that they give the same results as dense inputs. Please review when you have the time.

arjoly commented:
First thanks @jmschrei to give love to those codes. Here a few thoughts about this pull request.

I find great that you add sparsity support to gbrt. (Though I would have done that in another pr). As pointed by some people in the @ppret pull request, it is needed to store the sparse `X` matrix in both csr and csc format to handle sparse matrix efficiently. It allows to avoid costly format conversions between the csr and the csc format.

I understand the need and the benefits to merge more of the logic between pre-sort best splitter and the best spliter. In this pr, it is proposed to interleave the codes with `if else`. If we go to more rationalization such as merging the random splitter and the sparse splitters, this approach might start leading to spaghetti code. What do you think of having inline methods which would be overriden by each concrete implementation? The advantage is that it would add code re-use with some semantics. 

I am not fan of adding a `presort` parameter to the random forest estimators. Since that most (all?) people builds totally developed trees, it might not be used much and add some complexify in terms of usage and maintenance. In the gradient boosting case, I wouldn't allow the user to choose and switch presort or not depending on the matrix sparsity.

jmschrei commented:
Thanks for the comments @arjoly.

To be clear, I haven't added in sparsity support for presorting. I added an option to turn presorting off for gradient boosting, and when you do, you get the same sparsity support that decision trees/random forests get. Thus not having a separate PR for it.

I can understand the concern about spaghetti code.  Presorting and best splitting are so similar that merging them makes much more sense than any other merger, because it's not an algorithmic difference in how the split is calculated, just in how Xf is prepared for the split. This requires only the addition of 31 lines of code (15 of which are initializing and declaring variables) versus having an entirely new object and ~250 lines of repeated code.

Having inline methods may be the correct way to merge more code in the future. I was thinking about factorizing the node_split method into several methods, such as "sort_array", "sort_array_w_presorting", and "sort_sparse_array", which would return Xf. Then have a "best_split" method which takes in Xf and finds the best split, or a "random_split" method which finds the best random split. Lastly, a "partition_samples" method which is shared. I haven't fully formed the idea yet, but if you have ideas I'd love to see them.

I like the addition of presorting for these other methods, but am not tied to it--it seems like a "why not" case. It's easy to do add in, the default behaviour is not to use it so users won't notice a difference unless they know what they're doing, and should it become difficult to maintain it can be easily removed. Maybe @glouppe  will have an opinion.

glouppe commented:
> I am not fan of adding a presort parameter to the random forest estimators. Since that most (all?) people builds totally developed trees, it might not be used much and add some complexify in terms of usage and maintenance. In the gradient boosting case, I wouldn't allow the user to choose and switch presort or not depending on the matrix sparsity.

It is true that presorting is not very useful outside of GBRT. +1 for not adding this parameter in forests and keep the public API as it is.

jmschrei commented:
Presorting has been removed as an option for tests. Code, unit tests, and documentation have been updated to reflect this.

glouppe commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -934,6 +935,13 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             snapshoting.
 
+        presort : bool, optional (default=False)
+
```
remove this empty line

glouppe commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1318,6 +1330,14 @@ class GradientBoostingClassifier(BaseGradientBoosting, ClassifierMixin):
         by `np.random`.
 
+    presort : bool, optional (default=False)
+
```
same here

glouppe commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1644,6 +1666,13 @@ class GradientBoostingRegressor(BaseGradientBoosting, RegressorMixin):
         by `np.random`.
 
+    presort : bool, optional (default=False)
+
```
and here

glouppe commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -993,22 +1016,6 @@ def test_non_uniform_weights_toy_edge_case_reg():
         assert_greater(gb.predict([[1, 0]])[0], 0.5)
 
 
-def test_non_uniform_weights_toy_min_weight_leaf():
```
Why is this test removed?

jmschrei replied:
It is removed because for trees built with gradient boosting in master, `tree.splitter` is a splitter object, but it's a string for all other trees built (and there are unit tests ensuring this is the desired behavior.) Since 'presort' is now an option, as opposed to indicated by passing in a hardcoded splitter object, `tree.splitter` is a string in gradient boosting. The call `tree.splitter.min_weight_leaf` thus doesn't hold anymore.

glouppe commented on sklearn/tree/_splitter.pyx:
```diff
@@ -257,6 +276,17 @@ cdef class BaseDenseSplitter(Splitter):
+
+            self.n_total_samples = X.shape[0]
+            self.sample_mask = <SIZE_t*> calloc(self.n_total_samples, 
+                sizeof(SIZE_t))
```
You should use `safe_realloc` here instead

glouppe commented on sklearn/tree/tree.py:
```diff
@@ -52,7 +52,7 @@
 
 DENSE_SPLITTERS = {"best": _splitter.BestSplitter,
-                   "presort-best": _splitter.PresortBestSplitter,
+                   "presort-best": _splitter.BestSplitter,
```
This is now redundant with `"best": _splitter.BestSplitter`

glouppe replied:
You should defined instead `"presort-best": partial(_splitter.BestSplitter, presort=True)`. Though I am not sure it would work with your API.

glouppe commented:
You should tests for presorting in `test_tree.py`. You can do that I believe by changing

```
"Presort-DecisionTreeClassifier": partial(DecisionTreeClassifier, splitter="presort-best"),
```

to 

```
"Presort-DecisionTreeClassifier": partial(DecisionTreeClassifier, presort=True),
```

(and the same for the regressor)

jmschrei commented:
Thanks for the review @glouppe! Good catch, I forgot to change that over. I've incorporated all your comments. It looks like a bunch of TSNE unit tests are failing. :(

glouppe commented:
Great, thanks for the changes! +1 for merge on my side.

glouppe commented:
@jmschrei It seems there are conflicts in your branch? (I imagine this comes from the what's new)

@arjoly Any more comments?

jmschrei commented:
Yep, rebased this branch onto master.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -934,6 +935,12 @@ def fit(self, X, y, sample_weight=None, monitor=None):
             computing held-out estimates, early stopping, model introspect, and
             snapshoting.
 
+        presort : bool, optional (default=False)
```
Should it be true by default?
Is there any benefits in letting the user choose?

arjoly replied:
Should we provide an auto mode (to handle sparse data)?

arjoly commented:
>I find great that you add sparsity support to gbrt. (Though I would have done that in another pr). As pointed by some people in the @ppret pull request, it is needed to store the sparse X matrix in both csr and csc format to handle sparse matrix efficiently. It allows to avoid costly format conversions between the csr and the csc format.

This needs to be handled.

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +277,10 @@ def fit(self, X, y, sample_weight=None, check_input=True):
 
+        if X_idx_sorted is None and self.presort == True and not issparse(X):
+            X_idx_sorted = np.asfortranarray(np.argsort(X, axis=0),
+                                             dtype=np.int32)
```
What is the benefit of sorting here instead of making this in the splitter?
If it's inside the splitter, it seems to remove a lot of complexity ovreall.

arjoly replied:
Ok, I get  the idea is to have "stateless" splitter for multi-threading.
This should be noted somewhere.

arjoly replied:
@jmschrei Can you confirm?

GaelVaroquaux replied:
> Ok, I get the idea is to have "stateless" splitter for multi-threading.
> This should be noted somewhere.

Indeed, I think a note on the code would help future people understanding
the design.

jmschrei replied:
It is to make it stateless, and also make it cleaner what is happening. Having a splitter object which remembers a dataset is more difficult to understand than passing in a presorted matrix. It also allowed for parallel trees to be built without a duplication of effort when I thought forests might want to have presorting. I have made a note.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1017,11 +1029,7 @@ def _fit_stages(self, X, y, y_pred, sample_weight, random_state,
-                                       self.min_samples_leaf,
-                                       min_weight_leaf,
-                                       random_state)
+        splitter = 'best'
```
this line is not needed anymore

arjoly commented:
I would add a new private method to the tree learner `_validate_X_fit` to share all input checking line. It would be nice if it would return optionally X_sorted.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -45,7 +45,6 @@
 from ..tree.tree import DecisionTreeRegressor
 from ..tree._tree import DTYPE, TREE_LEAF
-from ..tree._splitter import PresortBestSplitter
 from ..tree._criterion import FriedmanMSE
```
Is this import still needed?

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -48,22 +55,23 @@
-            "Train deviance does not monotonically decrease."
+            deviance_decrease = (clf.train_score_[:-1] - clf.train_score_[1:])
+            assert np.any(deviance_decrease >= 0.0), \
+                "Train deviance does not monotonically decrease."
```
It would be nice to use `assert_True`

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
-        assert error_rate < 0.08, ("Stochastic GradientBoostingClassifier(loss={}) "
-                                   "failed with error {}".format(loss, error_rate))
+    for presort in True, False:
+        for loss in ('deviance', 'exponential'):
```
It would be nice to yield the test with the set of params.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
+            gbrt.fit(X_train, y_train)
+            error_rate = (1.0 - gbrt.score(X_test, y_test))
+            assert error_rate < 0.09, \
+                "GB(loss={}) failed with error {}".format(loss, error_rate)
```
It would be nice to use assert_less.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
+            gbrt.fit(X_train, y_train)
+            error_rate = (1.0 - gbrt.score(X_test, y_test))
+            assert error_rate < 0.08, ("Stochastic GradientBoostingClassifier(loss={}) "
+                                       "failed with error {}".format(loss, error_rate))
```
It would be nice to use assert_less.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
-                last_y_pred = y_pred
+    for presort in True, False:
+        for loss in ("ls", "lad", "huber"):
+            for subsample in (1.0, 0.5):
```
It would be nice to yield the test with the set of params.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
-                "and score = %f" % (subsample, score)
+    for presort in True, False:
+        for subsample in (1.0, 0.5):
+            for sample_weight in (None, np.ones(len(iris.target))):
```
It would be nice to yield the test with the set of params.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
+                                                   n_classes=20)
+    y = y[:, 0]
+
+    for name, sparse_matrix in product([GradientBoostingClassifier,
```
name is an EstimatorClass

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
         gb.fit(X, y, sample_weight=sample_weight)
         assert_array_equal(gb.predict([[1, 0]]), [1])
 
+def check_sparse_input(clf, X, X_sparse, y):
```
clf is an EstimatorClass

arjoly replied:
I would change clf to EstimatorClass to avoid any confusion.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
+
+    assert_array_almost_equal(sparse.apply(X), dense.apply(X))
+
+    if isinstance(dense, GradientBoostingRegressor):
```
I don't understand this.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
+    if isinstance(dense, GradientBoostingRegressor):
+        assert_array_almost_equal(sparse.predict(X), dense.predict(X))
+        assert_array_almost_equal(sparse.feature_importances_,
+                                  dense.feature_importances_)
```
feature importance should be tested in classification and regression.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
+    assert_array_almost_equal(sparse.apply(X), dense.apply(X))
+
+    if isinstance(dense, GradientBoostingRegressor):
+        assert_array_almost_equal(sparse.predict(X), dense.predict(X))
```
This should be tested in classification and regression.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1029,34 @@ def test_non_uniform_weights_toy_edge_case_clf():
 
+def check_sparse_input(clf, X, X_sparse, y):
+    dense = clf(n_estimators=10, random_state=0, max_depth=2, 
+                presort=True).fit(X, y)
```
I wouldn't set the presort agurment here.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -227,48 +247,59 @@ def test_regression_synthetic():
+                                        presort=presort)
+        clf.fit(X, y)
+        #feature_importances = clf.feature_importances_
+        assert_true(hasattr(clf, 'feature_importances_'))
```
Why is it changed? `feature_importances_ ` is a property so it should always be true?

jmschrei replied:
This isn't changed. I'm just keeping it as part of the loop as a failsafe, because the later tests in the loop are important.

arjoly replied:
the commented code `#feature_importances = clf.feature_importances_` is strange.
I would remove the line as it induces me an error.

jmschrei replied:
Okay. I had let it be, in case it was a note left by another contributor for themselves for the future. I have removed it, and other commented code.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -4,6 +4,13 @@
 
+from itertools import product
+
+from scipy.misc import comb
```
Is it used?

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
+                    y_pred = clf.predict(boston.data)
+                    mse = mean_squared_error(boston.target, y_pred)
+                    assert mse < 6.0, "Failed with loss %s and " \
+                        "mse = %.4f" % (loss, mse)
```
It would be nice to use assert_less.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,85 @@ def test_classification_synthetic():
+                        "mse = %.4f" % (loss, mse)
+
+                    if last_y_pred is not None:
+                        np.testing.assert_array_almost_equal(
```
It would be nice to re-use the assert from sklearn.utils.testing

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -233,18 +234,34 @@ cdef class BaseDenseSplitter(Splitter):
+        self.X_idx_sorted_ptr = NULL
+        self.X_idx_sorted_stride = 0
+        self.sample_mask = NULL
+        self.presort = presort
```
Setting presort should raise an error for splitters that doesn't support presort.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -257,6 +274,16 @@ cdef class BaseDenseSplitter(Splitter):
+            self.X_idx_sorted = X_idx_sorted
+            self.X_idx_sorted_ptr = <INT32_t*> self.X_idx_sorted.data
+            self.X_idx_sorted_stride = (<SIZE_t> self.X_idx_sorted.strides[1] /
+                                       <SIZE_t> self.X_idx_sorted.itemsize)
```
There is ones psace missing here.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -351,18 +386,28 @@ cdef class BestSplitter(BaseDenseSplitter):
+                else:
+                    for i in range(start, end):
+                        Xf[i] = X[self.X_sample_stride * samples[i] +
+                                self.X_fx_stride * current.feature]
```
some spaces are missing here.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -351,18 +386,28 @@ cdef class BestSplitter(BaseDenseSplitter):
+                        if sample_mask[j] == 1:
+                            samples[p] = j
+                            Xf[p] = X[self.X_sample_stride * j +
+                                      self.X_fx_stride * current.feature]
```
feature offset could be re-used here.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -257,6 +274,16 @@ cdef class BaseDenseSplitter(Splitter):
 
+        if self.presort == 1:
+            self.X_idx_sorted = X_idx_sorted
+            self.X_idx_sorted_ptr = <INT32_t*> self.X_idx_sorted.data
```
Do we expect a different stride than X?

arjoly commented:
Setting presort with the random splitter should raise an error.

jmschrei commented:
Comments have been incorporated. Thanks for the thorough review @arjoly. 

I made an error with the criterion cleanup PR in that I forgot to recompile the criterion file, causing the .c file to be correct, but the .pyx to include a few variable name errors. I have fixed those in addition here. Rebasing also seems to have caused some issues, so I manually re-did those changes.

arjoly commented:
There are still modification in `forest.py`. Is it wanted?

arjoly commented:
I think that we need add an auto mode for the presort parameter, which select the splitter depending on the data sparsity. It would so nice from a usability point of view.

>I find great that you add sparsity support to gbrt. (Though I would have done that in another pr). As pointed by some people in the @ppret pull request, it is needed to store the sparse X matrix in both csr and csc format to handle sparse matrix efficiently. It allows to avoid costly format conversions between the csr and the csc format.

This comment still needs to be addressed.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -148,73 +156,88 @@ def test_classification_synthetic():
+    last_y_pred = None
+    for i, sample_weight in enumerate(
+            (None, np.ones(len(boston.target)),
+             2 * np.ones(len(boston.target)))):
```
nitpick: Strange indentation.

I would do 
```
or i, sample_weight in enumerate((..., 
                                  ...,
                                  ...)):
```

or define the list of sample weight before the loop.

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -907,13 +931,13 @@ def test_zero_estimator_clf():
     est.fit(X, y)
 
-    assert est.score(X, y) > 0.96
+    assert_greater(est.score(X, y), 0.96)
```
Thanks !

arjoly commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -1022,3 +1030,32 @@ def test_non_uniform_weights_toy_edge_case_clf():
+
+
+def test_sparse_input():
+    clfs = (GradientBoostingClassifier, GradientBoostingRegressor)
```
It would rename clfs to EstimatorClasses.

arjoly replied:
clf is usually used as the abbreviation of a classifier.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -351,18 +389,28 @@ cdef class BestSplitter(BaseDenseSplitter):
+                        if sample_mask[j] == 1:
+                            samples[p] = j
+                            Xf[p] = X[self.X_sample_stride * j +
+                                      self.X_feature_stride * current.feature]
```
`feature_offset = self.X_idx_sorted_stride * current.feature` is it equal to  `self.X_feature_stride * current.feature`?
If yes, I would re-use feature offset.

jmschrei replied:
Unfortunately can't be assumed to be equal

arjoly replied:
Could you explain? Otherwise, you can do once the computation of `self.X_feature_stride * current.feature`.

jmschrei replied:
See other line note.

I tried to get rid of one, but unit tests do fail when I do such.

arjoly replied:
`self.X_feature_stride * current.feature` should be computed once.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -255,7 +274,18 @@ cdef class BaseDenseSplitter(Splitter):
+            self.X_idx_sorted = X_idx_sorted
+            self.X_idx_sorted_ptr = <INT32_t*> self.X_idx_sorted.data
+            self.X_idx_sorted_stride = (<SIZE_t> self.X_idx_sorted.strides[1] /
+                                        <SIZE_t> self.X_idx_sorted.itemsize)
```
Do you need this stride?

jmschrei replied:
Yes. Usually X_sample_stride is the number of features, X_feature_stride is 1, and X_idx_sorted_stride is the number of samples. However, if a transposed matrix is passed in, then X_feature_stride and X_sample_stride are reversed.

arjoly replied:
Don't we already these information with the X_sample_stride and X_feature_stride?

glouppe replied:
A priori I dont think we have guarantees that X_idx_sorted is stored in the same way as X itself.

arjoly replied:
thanks

jmschrei replied:
If we have a matrix with 10 samples and 5 features, X_feature_stride is 1 and X_sample_stride is 5. Feature stride is 1 because you only need to move over 1 to get to the next feature in the same sample. If you transpose this matrix, now X_sample_stride is 1, and X_feature_stride is 10. However, since you generate the index matrix from the appropriately oriented matrix, it's always has a feature stride of 10.

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -102,7 +103,8 @@ def __init__(self,
 
-    def fit(self, X, y, sample_weight=None, check_input=True):
+    def fit(self, X, y, sample_weight=None, check_input=True, 
+        X_idx_sorted=None):
```
The indentation is weird here.

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +277,15 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+        # but do not handle any presorting themselves. Ensemble algorithms which
+        # desire presorting must do presorting themselves and pass that matrix
+        # into each tree.
+        if X_idx_sorted is None and self.presort == True and not issparse(X):
```
If it is not None, I would check that the shape matches the X.

arjoly replied:
This behavior doesn't seems to be well represented in doctstring of presort. Have I missed something?

jmschrei commented:
>> I find great that you add sparsity support to gbrt. (Though I would have done that in another pr). As pointed by some people in the @ppret pull request, it is needed to store the sparse X matrix in both csr and csc format to handle sparse matrix efficiently. It allows to avoid costly format conversions between the csr and the csc format.

>This comment still needs to be addressed.

I don't understand what you're asking. Are you asking me to add a new feature? I am doing sparse splitting the same way that it was being done before.

jmschrei commented:
I've addressed comments, added an 'auto' default option, and added some more unit tests. Since this PR is getting rather hefty now, I am going to submit another PR in the future applying the same style changes to _splitter.pyx  as _criterion.pyx, instead of handling that here.

arjoly commented:
To grow a tree on sparse, you need the data to be in sparse csc format. However to make a prediction, you need a sparse csr data.This means that whenever you build a boosting models at each step you:
(1) convert the X matrix to csc if needed (2) build the tree (3) convert the X matrix to csr (4) predict with the current tree and (5) repeat (1) until we have done n_estimators. This transformation from csc and csr induce computational costs that should be avoided.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1318,6 +1317,12 @@ class GradientBoostingClassifier(BaseGradientBoosting, ClassifierMixin):
         If None, the random number generator is the RandomState instance used
         by `np.random`.
 
+    presort : bool, optional (default='auto')
```
It is `presort : bool or 'auto', optional (default='auto')`

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1644,6 +1651,11 @@ class GradientBoostingRegressor(BaseGradientBoosting, RegressorMixin):
         If None, the random number generator is the RandomState instance used
         by `np.random`.
 
+    presort : bool, optional (default='auto')
```
same here

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +277,29 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+        if self.presort == 'auto' and issparse(X):
+            self.presort = False
+        elif self.presort == 'auto':
+            self.presort = True
```
We shouldn't change the value of `self.presort`.

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +277,29 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+            X_idx_sorted = np.asfortranarray(np.argsort(X, axis=0),
+                                             dtype=np.int32)
+        elif X_idx_sorted is not None:
+            if X_idx_sorted.shape != X.shape:
```
Both if could be combined.

arjoly commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1303,3 +1303,21 @@ def test_public_apply():
+
+def test_presort_sparse():
+    ests = (partial(DecisionTreeClassifier, presort=True)(), 
+            partial(DecisionTreeRegressor, presort=True)())
```
Nitpick: we don't need the partial.

jmschrei replied:
I usually prefer to be explicit as opposed to implicit in the case of unit tests, so that it is clear what is being tested. What are your thoughts?

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +277,29 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+        elif X_idx_sorted is not None:
+            if X_idx_sorted.shape != X.shape:
+                raise ValueError("Shape of presorted matrix does not match "
+                                 "shape of data matrix.")
```
I would say something like
```
ValueError("The shape of X (X.shape = {}) doesn't match the shape of 
           "X_idx_sorted (X_idx_sorted.shape={})."
           "".format(X.shape, X_idx_sorted.shape))
```

jmschrei commented:
> To grow a tree on sparse, you need the data to be in sparse csc format. However to make a prediction, you need a sparse csr data.This means that whenever you build a boosting models at each step you:
(1) convert the X matrix to csc if needed (2) build the tree (3) convert the X matrix to csr (4) predict with the current tree and (5) repeat (1) until we have done n_estimators. This transformation from csc and csr induce computational costs that should be avoided.

Got it. Apologies for the confusion.

jmschrei commented:
Comments have been incorporated. For sparse data, csc matrices are passed into fitting functions, and csr matrices are passed into prediction functions. Thanks @arjoly for the thorough review.

glouppe commented:
Any more comments @arjoly ? If not I believe we can merge this one.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -728,11 +740,12 @@ def __init__(self, loss, learning_rate, n_estimators, min_samples_split,
 
     def _fit_stage(self, i, X, y, y_pred, sample_weight, sample_mask,
-                   criterion, splitter, random_state):
+                   random_state, X_idx_sorted, X_csc=None, X_csr=None):
```
Here it would be better to have X_fit and X_predict instead of X, X_csc and X_csr.

jmschrei replied:
I disagree. X_csc and X_csr is explicit about what those are. If I saw X_fit and X_predict in the wild, I'd assume that they were different datasets. Here they are the same dataset.

arjoly commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -981,9 +1004,24 @@ def fit(self, X, y, sample_weight=None, monitor=None):
+            presort = True
+        
+        if presort == True and issparse(X):
+            if issparse(X):
```
This is not necessary given line above.

arjoly replied:
This could be indeed simplify
```
if presort == True: 
    if issparse(X):
        ....
    else:
       ...
```

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -351,18 +386,28 @@ cdef class BestSplitter(BaseDenseSplitter):
+                else:
+                    for i in range(start, end):
+                        Xf[i] = X[self.X_sample_stride * samples[i] +
+                                  self.X_feature_stride * current.feature]
```
`self.X_feature_stride * current.feature` should be computed once.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -423,7 +468,7 @@ cdef class BestSplitter(BaseDenseSplitter):
             while p < partition_end:
                 if X[X_sample_stride * samples[p] +
-                     X_fx_stride * best.feature] <= best.threshold:
+                     X_feature_stride * best.feature] <= best.threshold:
```
`X_feature_stride * best.feature` could be computed once.

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -663,13 +714,13 @@ cdef class RandomSplitter(BaseDenseSplitter):
                 # Find min, max
                 min_feature_value = X[X_sample_stride * samples[start] +
-                                      X_fx_stride * current.feature]
+                                      X_feature_stride * current.feature]
```
` X_feature_stride * current.feature` could be computed once

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -663,13 +714,13 @@ cdef class RandomSplitter(BaseDenseSplitter):
                 for p in range(start + 1, end):
                     current_feature_value = X[X_sample_stride * samples[p] +
-                                              X_fx_stride * current.feature]
+                                              X_feature_stride * current.feature]
```
` X_feature_stride * current.feature` could be computed once

arjoly commented on sklearn/tree/_splitter.pyx:
```diff
@@ -743,7 +794,7 @@ cdef class RandomSplitter(BaseDenseSplitter):
                 while p < partition_end:
                     if X[X_sample_stride * samples[p] +
-                         X_fx_stride * best.feature] <= best.threshold:
+                         X_feature_stride * best.feature] <= best.threshold:
```
` X_feature_stride * best.feature` could be computed once

arjoly commented on sklearn/tree/tree.py:
```diff
@@ -269,6 +284,33 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+            X_idx_sorted = np.asfortranarray(np.argsort(X, axis=0),
+                                             dtype=np.int32)
+
+            if not presort and X_idx_sorted.shape != X.shape:
```
Is this line ever reach?

arjoly commented:
I made a last round of review.

jmschrei commented:
Comments have been into account, and commits have been squashed.

arjoly commented:
LGTM

arjoly commented:
thanks @jmschrei

GaelVaroquaux commented:
> thanks @jmschrei

Great job once again!

jmschrei commented:
:trumpet:
